src/vos/vos_query.c

```diff
@@ -203,25 +198,58 @@ query_recx(struct open_query *query, daos_recx_t *recx)
 	if (rc != 0)
 		goto fini;
 
-	recx->rx_idx = entry.en_sel_ext.ex_lo;
-	recx->rx_nr = entry.en_sel_ext.ex_hi - entry.en_sel_ext.ex_lo + 1;
+	if (opc & EVT_ITER_SKIP_HOLES) {
+		if (entry.en_visibility & EVT_VISIBLE) {
+			recx->rx_idx = entry.en_sel_ext.ex_lo;
+			recx->rx_nr = entry.en_sel_ext.ex_hi - entry.en_sel_ext.ex_lo + 1;
+		} else {
+			rc = -DER_NONEXIST;
+		}
+		D_GOTO(fini, rc);
+	}
+
+	/* The following only supports MAX recx query at the moment */
+	if (!(query->qt_flags & VOS_GET_MAX))
+		D_GOTO(fini, rc);
+
+	/* let's find the last punched extent */
+	while (!bio_addr_is_hole(&entry.en_addr)) {
```

Comment:
Why do you need the final punch if there is data after it?

Also, it would probably be a 1-2 line change in evtree sort algorithm to just show visible punches...something like 
EVT_ITER_SKIP_DATA

If it would help here.

Reply:
oh, this is suppose to find the last/largest punched extent, and I do not want put the size checking logic here.